packages/server/src/api/controllers/row/ExternalRequest.ts

```diff
@@ -260,10 +286,22 @@ export class ExternalRequest<T extends Operation> {
     }
   }
 
+  // seeds the object with table and datasource information
+  async retrieveMetadata(datasourceId: string) {
+    if (!this.datasource) {
+      this.datasource = await sdk.datasources.get(datasourceId)
+      if (!this.datasource || !this.datasource.entities) {
+        throw "No tables found, fetch tables before query."
+      }
+      this.tables = this.datasource.entities
+      this.tableList = Object.values(this.tables)
+    }
+  }
```

Comment:
It's a shame that this removes the type narrowing on this.datasource. Is there any way to retain that? For example, could we return the datasource from this function and assign it to this.datasource at the calling location? Because right now you've had to add an exclamation mark to the end of this.datasource at the call site.